---
title: "SNPs data preprocessing"
author: "Shuai, Yi"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
# Add required packages
  library(ggplot2)
  library(tidyverse)
  library(hrbrthemes)
  library("stringr")
  library(ComplexHeatmap)
```

#------------------------------------------------------
# 1. Preprocessing and basic stats of SNPs and CNV data 
#------------------------------------------------------

```{r}
# Preprocess the SNPs file
if (file.exists("./data/0_raw/2_raw.snps.table.rds")) {
  message("SNPs data merging, DONE!")
} else {
  files.snp = list.files("./data/0_raw/1_snp.cds_annot/")
  n.samples = length(files.snp)
  final.snps.table = NULL
  dbSNP_ver = NULL
  for (i in 1:n.samples) {
    temp.id = sapply(strsplit(x = files.snp[i], split = "\\."), `[`, 1)
    temp = read.csv(paste0("./data/0_raw/1_snp.cds_annot/",files.snp[i]))
    dbSNP_ver = c(dbSNP_ver, colnames(temp)[10])
    colnames(temp)[10] = "dbSNP_database"
    temp$sample_id = temp.id
    final.snps.table = rbind(final.snps.table, temp)
  }
  saveRDS(final.snps.table, file = "./data/0_raw/2_raw.snps.table.rds")
  saveRDS(dbSNP_ver, file = "./data/0_raw/2_raw.snps.dbSNPver.rds")
}

# Preprocess the CNVs file
if (file.exists("./data/0_raw/2_raw.cnv.table.rds")) {
  message("CNAs data merging, DONE!")
} else {
  files.cnv = list.files("./data/0_raw/1_indel.cds_annot/")
  n.samples = length(files.cnv)
  final.cnv.table = NULL
  dbSNP_ver = NULL
  for (i in 1:n.samples) {
    temp.id = sapply(strsplit(x = files.cnv[i], split = "\\."), `[`, 1)
    temp = read.csv(paste0("./data/0_raw/1_indel.cds_annot/",files.cnv[i]))
    dbSNP_ver = c(dbSNP_ver, colnames(temp)[10])
    colnames(temp)[10] = "dbSNP_database"
    temp$sample_id = temp.id
    final.cnv.table = rbind(final.cnv.table, temp)
  }
  saveRDS(final.cnv.table, file = "./data/0_raw/2_raw.cnv.table.rds")
  saveRDS(dbSNP_ver, file = "./data/0_raw/2_raw.cnv.dbSNPver.rds")
}
```

#----------------------------------------------------
# 2. Visualize basic stats of SNPs and clean the data
#----------------------------------------------------

```{r fig.height=4, fig.width=7}
# Read the raw data
  raw.snps.table = readRDS("./data/0_raw/2_raw.snps.table.rds")
  # raw.snps.table = final.snps.table
# 1. filter based on population frequency  
  # write.csv(raw.snps.table[1:6,], file = "./1_examplary_table.csv")
  ExAC_EAS_Freq = as.numeric(raw.snps.table$ExAC_EAS)
  X1000G_EAS_Freq=as.numeric(raw.snps.table$X1000G_EAS)
  sum(is.na(ExAC_EAS_Freq)) 
  sum(is.na(raw.snps.table$X1000G_EAS))
  ExAC_EAS_Freq[is.na(ExAC_EAS_Freq)] = 0
  X1000G_EAS_Freq[is.na(X1000G_EAS_Freq)] = 0
 
  data <- data.frame(ExAC_EAS = ExAC_EAS_Freq, X1000G_EAS = X1000G_EAS_Freq)

  p1 <- data %>%
    ggplot( aes(x=X1000G_EAS)) +
      geom_histogram(binwidth=0.002,  fill="#69b3a2", alpha=1) +
      ggtitle("Population frequency in X1000G_EAS") +
      theme_classic() +
      theme(
        plot.title = element_text(size=15)
      ) + geom_vline(xintercept = 0.01, linetype="dotted", color = "red", size=1)
  p1
  
  p2 <- data %>%
    ggplot( aes(x=ExAC_EAS)) +
      geom_histogram(binwidth=0.002, fill="#69b3a2", alpha=1) +
      ggtitle("Population frequency in ExAC_EAS") +
      theme_classic() +
      theme(
        plot.title = element_text(size=15)
      ) + geom_vline(xintercept = 0.01, linetype="dotted", color = "red", size=1)
  p2

  ggsave(plot = p1, filename = "./image/s1_data_qc/1_x1000g_eas.pdf", dpi = 300, width = 7, height = 4) 
  ggsave(plot = p2, filename = "./image/s1_data_qc/1_exac_eas.pdf", dpi = 300, width = 7, height = 4) 
```

```{r}
# 1. filter based on population frequency [Cont'] [81,301]
  raw.snps.table_qc1 = raw.snps.table[which(ExAC_EAS_Freq <= 0.01 & X1000G_EAS_Freq <= 0.01),]
  dim(raw.snps.table_qc1)

# 2. filter synonymous_variants [51,266]
  raw.snps.table_qc2 = raw.snps.table_qc1[-which(raw.snps.table_qc1$Func == "synonymous_variant"),]
  dim(raw.snps.table_qc2)
  
# 3. filter based on predicted pathogenic scores (SIFT, Polyphen2, and MutationTaster) [15,575]
  table(str_sub(raw.snps.table_qc2$SIFT,-2,-2))
  SIFT_re = str_replace_all(raw.snps.table_qc2$SIFT,"\\.,", "")
  SIFT_re = str_replace_all(SIFT_re,",\\.", "")
  # SIFT_re[73]
  table(str_sub(SIFT_re,-2,-2))
  raw.snps.table_qc2$SIFT_re = SIFT_re
  table(str_sub(raw.snps.table_qc2$PolyPhen2,-2,-2))
  table(str_sub(raw.snps.table_qc2$MutationTaster,-2,-2))
  
  raw.snps.table_qc2$SIFT_bi = str_sub(raw.snps.table_qc2$SIFT_re,-2,-2)
  raw.snps.table_qc2$PolyPhen2_bi = str_sub(raw.snps.table_qc2$PolyPhen2,-2,-2)
  raw.snps.table_qc2$MutationTaster_bi = str_sub(raw.snps.table_qc2$MutationTaster,-2,-2)
  # SIFT: Tolerated vs Damaging
  # Plolyphen2: benign, possibly damaging, or probably damaging; If the lack of data, unknown.
  # MutationTaster: disease causing; disease causing automatic; polymorphism; polymorphism automatic
  length(which(raw.snps.table_qc2$SIFT_bi == "T"  |
               raw.snps.table_qc2$PolyPhen2_bi == "B"|
               raw.snps.table_qc2$MutationTaster == "N"  |
               raw.snps.table_qc2$MutationTaster == "P" ))
  snps_benign = which(raw.snps.table_qc2$SIFT_bi == "T"| raw.snps.table_qc2$PolyPhen2_bi == "B"|
                      raw.snps.table_qc2$MutationTaster_bi == "N"| raw.snps.table_qc2$MutationTaster_bi == "P" )
  raw.snps.table_qc3 = raw.snps.table_qc2[-snps_benign,]
  dim(raw.snps.table_qc3)
  table(raw.snps.table_qc3$Func)

# 4. filter based on impact [15,556]
  raw.snps.table_qc4 = raw.snps.table_qc3[-which(raw.snps.table_qc3$Impact == "LOW"),]
  table(raw.snps.table_qc4$Func)
  
  raw.snps.table_qc4$Chr = factor(raw.snps.table_qc4$Chr, c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9",
                                                            "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17",
                                                            "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"))
  saveRDS(raw.snps.table_qc4, file = "./data/1_final_SNVs_table.rds")
```

#----------------------------------------------------
# 3. Visualize basic stats of CNAs and clean the data
#----------------------------------------------------

```{r}
# Read the raw data [55,099]
  raw.cnas.table = readRDS("./data/0_raw/2_raw.cnv.table.rds")
  ExAC_EAS_Freq = as.numeric(raw.cnas.table$ExAC_EAS)
  X1000G_EAS_Freq=as.numeric(raw.cnas.table$X1000G_EAS)
  ExAC_EAS_Freq[is.na(ExAC_EAS_Freq)] = 0
  X1000G_EAS_Freq[is.na(X1000G_EAS_Freq)] = 0

# 1. filter based on population frequency [Cont'] [9,548]
  raw.cnas.table_qc1 = raw.cnas.table[which(ExAC_EAS_Freq <= 0.01 & X1000G_EAS_Freq <= 0.01),]
  dim(raw.cnas.table_qc1)
# 2. filter indels located at intron [7,526]
  intron_indel =  which(raw.cnas.table_qc1$Func == "splice_acceptor_variant+intron_variant" |                                                                       
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_region_variant+conservative_inframe_deletion+intron_variant" |               
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_donor_variant+intron_variant" |                                                  
                        raw.cnas.table_qc1$Func == "splice_donor_variant+intron_variant"                           |                                                
                        raw.cnas.table_qc1$Func == "splice_donor_variant+splice_region_variant+intron_variant"       |                                              
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_region_variant+intron_variant"      |                                            
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_donor_variant+splice_region_variant+disruptive_inframe_deletion+intron_variant" |
                        raw.cnas.table_qc1$Func == "frameshift_variant+splice_acceptor_variant+splice_region_variant+intron_variant"                               |
                        raw.cnas.table_qc1$Func == "frameshift_variant+splice_donor_variant+splice_region_variant+intron_variant"                                  |
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+5_prime_UTR_variant+intron_variant"                                                    |
                        raw.cnas.table_qc1$Func == "splice_donor_variant+splice_region_variant+conservative_inframe_deletion+intron_variant"                       |
                        raw.cnas.table_qc1$Func == "frameshift_variant+splice_acceptor_variant+splice_donor_variant+splice_region_variant+intron_variant"          |
                        raw.cnas.table_qc1$Func == "splice_donor_variant+splice_region_variant+disruptive_inframe_deletion+intron_variant"                         |
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_region_variant+intron_variant+non_coding_transcript_exon_variant"               |
                        raw.cnas.table_qc1$Func == "splice_donor_variant+5_prime_UTR_variant+intron_variant"                                                       |
                        raw.cnas.table_qc1$Func == "splice_donor_variant+splice_region_variant+5_prime_UTR_variant+intron_variant"                                 |
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_region_variant+5_prime_UTR_variant+intron_variant"                              |
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_region_variant+disruptive_inframe_deletion+intron_variant"                      |
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+splice_donor_variant+splice_region_variant+intron_variant"| 
                        raw.cnas.table_qc1$Func == "splice_acceptor_variant+3_prime_UTR_variant+intron_variant")
  raw.cnas.table_qc2 = raw.cnas.table_qc1[-intron_indel,]
  dim(raw.cnas.table_qc2)
  unique(raw.cnas.table_qc2$Func)
  unique(raw.cnas.table_qc2$Chr)
  
# 3. filter based on impact [7,171]  
  raw.cnas.table_qc2$Chr = factor(raw.cnas.table_qc2$Chr, c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9",
                                                           "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17",
                                                           "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"))
  saveRDS(raw.cnas.table_qc2, file = "./data/1_final_CNAs_table.rds")
```

#--------------------
# 4. Chr distribution
#--------------------

```{r}
# 1. Prepare the input data  
  final.snps.table = readRDS("./data/1_final_SNVs_table.rds")
  final.cnas.table = readRDS("./data/1_final_CNAs_table.rds")
  chr_lenght = c(247249719, 242951149, 199501827, 191273063, 180857866, 170899992, 158821424, 146274826, 
                 140273252, 135374737, 134452384, 132349534, 114142980, 106368585, 100338915, 88827254, 
                 78774742, 76117153, 63811651, 62435964, 46944323, 49691432, 154913754, 57772954)
  chr_lenght_rl = chr_lenght/sum(chr_lenght) * 100
  data = data.frame(snps_counts = table(final.snps.table$Chr),
                    cnas_counts = table(final.cnas.table$Chr))
  data = data[,c(2,4)]
  colnames(data) = c("snps_counts", "cnas_counts")
  data$snps_adj = data$snps_counts/chr_lenght_rl
  data$cnas_adj = data$cnas_counts/chr_lenght_rl
  data$chr =c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9",
              "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17",
              "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY")
  data$chr = factor( data$chr, c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9",
                                 "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17",
                                 "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"))
  
  p3 <- ggplot(data, aes(x = chr, y = snps_adj)) +
    geom_bar(stat='identity',  fill="#69b3a2" , width=0.75) +
    ggtitle("SNPs frequencies per Chr adjusted by length") +
    theme_classic() +
    theme(
      plot.title = element_text(size=15),
      axis.text.x = element_text(angle = 45, vjust = 0.5)
    ) +
    geom_hline(yintercept = 300, linetype="dotted", color = "red", size=1)
  p4 <- ggplot(data, aes(x = chr, y = cnas_adj)) +
    geom_bar(stat='identity',  fill="#69b3a2" , width=0.75) +
    ggtitle("Indels frequencies per Chr adjusted by length") +
    theme_classic() +
    theme(
      plot.title = element_text(size=15),
      axis.text.x = element_text(angle = 45, vjust = 0.5)
    ) +
    geom_hline(yintercept = 150, linetype="dotted", color = "red", size=1)
  p3
  p4

  ggsave(plot = p3, filename = "./image/1_snps_per_unit_chr.pdf", dpi = 300, width = 7, height = 4) 
  ggsave(plot = p4, filename = "./image/2_indels_per_unit_chr.pdf", dpi = 300, width = 7, height = 4) 
```

#-------------------------------------
# 5. Oncoprint plot of SNPs and Indels
#-------------------------------------

```{r fig.width=6, fig.height=6}
# 1. Load the SNVs data
  final.snps.table = readRDS("./data/1_final_SNVs_table.rds")
  clinical.table = read.csv(paste0("./data/0_list_processed.csv"))
  
  # add smoking info
  final.snps.table$smoking = 1
  for (i in 1:124) {
    if (clinical.table[i,"Smoking"] == 0) {
      final.snps.table$smoking[which(final.snps.table$sample_id == clinical.table[i,"patient_id"])] = 0
  }  }
  
  # table(final.snps.table$Func)
  # missense_variant; missense_variant+splice_region_variant; splice_acceptor_variant+intron_variant; splice_donor_variant+intron_variant 
  # start_lost; start_lost+splice_region_variant; stop_gained; stop_gained+splice_region_variant; stop_lost

  # table(final.snps.table$Func[which(final.snps.table$Exonic.Biotype == "MISSENSE")])
  # missense_variant; missense_variant+splice_region_variant; start_lost; start_lost+splice_region_variant; stop_lost

# 2. Rename the variants  
  final.snps.table$Func[which(final.snps.table$Func == "missense_variant"  |
                              final.snps.table$Func == "missense_variant+splice_region_variant" |
                              final.snps.table$Func == "start_lost+splice_region_variant"  |
                              final.snps.table$Func == "start_lost"|
                              final.snps.table$Func == "stop_lost" )] = "missense"
  final.snps.table$Func[which(final.snps.table$Func == "stop_gained")] = "nonsense"
  final.snps.table$Func[which(final.snps.table$Func == "splice_donor_variant+intron_variant"    |
                              final.snps.table$Func == "splice_acceptor_variant+intron_variant" |
                              final.snps.table$Func == "stop_gained+splice_region_variant")] = "splice_region"
  # table(final.snps.table$Func)
  # table(final.snps.table$Func)/sum(table(final.snps.table$Func))
  # mean(table(final.snps.table$sample_id))
  # sd(table(final.snps.table$sample_id))
  # sort(table(final.snps.table$Gene), decreasing = T)

# 3. Prepare the input data for High impact SNVs
  sample_gene_list = unique(paste0(final.snps.table$sample_id, "_", final.snps.table$Gene))
  recurrent_snv = names(which(table(sapply(strsplit(sample_gene_list, split = "_"), '[', 2)) >= 12))
  mat = matrix(data = "",nrow = length(recurrent_snv), ncol = 124)
  colnames(mat) = unique(final.snps.table$sample_id)
  rownames(mat) = recurrent_snv
  for (i in recurrent_snv) {
    for (j in unique(final.snps.table$sample_id)) {
      temp_id = which(final.snps.table$Gene == i & final.snps.table$sample_id == j)
      if (!identical(temp_id, integer(0))) {
        temp = final.snps.table[temp_id,]
        mat[i,j] = paste(temp$Func, collapse = ";")
      }
    }
  }

  col = c(missense = "#69b3a2", nonsense = "#E21818", splice_region = "#FFDD83")
  p5 = oncoPrint(mat, alter_fun_is_vectorized = FALSE,
                alter_fun = list(
                    missense =  function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                        gp = gpar(fill = col["missense"], col = NA)),
                    nonsense =  function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                        gp = gpar(fill = col["nonsense"], col = NA)),
                    splice_region = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                        gp = gpar(fill = col["splice_region"], col = NA)) ),
                col = col)
  pdf(file = "./image/3_snps_landscape.pdf", width = 6, height = 6)
  p5
  dev.off()
```

```{r}
# 1. Load the CNAs data
  final.cnas.table = readRDS("./data/1_final_CNAs_table.rds")

  final.cnas.table$Func[which(final.cnas.table$Func == "disruptive_inframe_deletion" | 
                              final.cnas.table$Func == "stop_lost+conservative_inframe_deletion" | 
                              final.cnas.table$Func == "stop_lost+disruptive_inframe_deletion" | 
                              final.cnas.table$Func == "start_lost+conservative_inframe_deletion" | 
                              final.cnas.table$Func == "stop_lost+splice_region_variant+conservative_inframe_deletion" | 
                              final.cnas.table$Func == "disruptive_inframe_deletion" )] = "deletion"
  
  final.cnas.table$Func[which(final.cnas.table$Func == "disruptive_inframe_insertion" | 
                              final.cnas.table$Func == "start_lost+conservative_inframe_insertion" | 
                              final.cnas.table$Func == "start_lost+disruptive_inframe_insertion" | 
                              final.cnas.table$Func == "stop_gained+disruptive_inframe_insertion" |
                              final.cnas.table$Func == "stop_gained+conservative_inframe_insertion" |
                              final.cnas.table$Func == "stop_gained+splice_region_variant+conservative_inframe_insertion")] = "insertion"
  
  final.cnas.table$Func[which(final.cnas.table$Func == "frameshift_variant" | 
                              final.cnas.table$Func == "frameshift_variant+stop_gained" | 
                              final.cnas.table$Func == "frameshift_variant+splice_region_variant" | 
                              final.cnas.table$Func == "frameshift_variant+start_lost+splice_region_variant" | 
                              final.cnas.table$Func == "frameshift_variant+stop_gained+splice_region_variant" | 
                              final.cnas.table$Func == "frameshift_variant+stop_lost" | 
                              final.cnas.table$Func == "frameshift_variant+start_lost")] = "frameshift"
  # table(final.cnas.table$Func)
  # table(final.cnas.table$Func)/sum(table(final.cnas.table$Func))
  # mean(table(final.cnas.table$sample_id))
  # sd(table(final.cnas.table$sample_id))

# 2. Prepare the input data for High impact SNVs # 124 * 0.3
  sample_gene_list = unique(paste0(final.cnas.table$sample_id, "_", final.cnas.table$Gene))
  recurrent_cna = names(which(table(sapply(strsplit(sample_gene_list, split = "_"), '[', 2)) >= 37))
  
  mat = matrix(data = "",nrow = length(recurrent_cna), ncol = 124)
  colnames(mat) = unique(final.cnas.table$sample_id)
  rownames(mat) = recurrent_cna
  for (i in recurrent_cna) {
    for (j in unique(final.cnas.table$sample_id)) {
      temp_id = which(final.cnas.table$Gene == i & final.cnas.table$sample_id == j)
      if (!identical(temp_id, integer(0))) {
        temp = final.cnas.table[temp_id,]
        mat[i,j] = paste(temp$Func, collapse = ";")
      }
    }
  }
# 3. Plot
  # unique(final.cnas.table$Func)
  col = c(deletion = "#69b3a2",
          insertion = "#E21818",
          frameshift = "#FFDD83")
  p6 = oncoPrint(mat, alter_fun_is_vectorized = FALSE,
        alter_fun = list(
            deletion =  function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                gp = gpar(fill = col["deletion"], col = NA)),
            insertion =  function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                gp = gpar(fill = col["insertion"], col = NA)),
            frameshift = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                gp = gpar(fill = col["frameshift"], col = NA)) ), col = col)
  pdf(file = "./image/4_cnas_landscape.pdf", width = 6, height = 6)
  p6
  dev.off()
```

#----------------------------------------------------
# 4. The signature of recurrent r.v. genes
#----------------------------------------------------

```{r fig.height=3, fig.width=10}
  final.snps.table = readRDS("./data/1_final_SNVs_table.rds")
  final.cnas.table = readRDS("./data/1_final_CNAs_table.rds")
  recurrent_vars = union(recurrent_cna, recurrent_snv)
  save(recurrent_vars, file = "./data/1_final_recur_genes.RData")
  load("./data/1_final_recur_genes.RData")
  library(clusterProfiler)  
  library(DOSE)
  library("org.Hs.eg.db")  
  ego <- enrichGO(gene          = recurrent_vars,
                  keyType       = 'SYMBOL',
                  universe      = union(final.snps.table$Gene, final.cnas.table$Gene),
                  OrgDb         = org.Hs.eg.db,
                  ont           = "ALL",
                  pAdjustMethod = "none", 
                  pvalueCutoff  = 0.05,
                  readable      = TRUE)
  ego@result
  save(ego, file = "./data/1_gsea_recur_genes.RData")
  
  load(file = "./data/1_gsea_recur_genes.RData")
  y <- as.data.frame(ego)
  enrichment_plot <- y[c(1:5,8:10,11,12,13:16),] %>% 
    arrange(p.adjust) %>%
    mutate(Description = fct_reorder(Description, p.adjust, .desc = T))
  p =
  ggplot(enrichment_plot, aes(x = Description, y = -log10(p.adjust), fill = p.adjust)) +
    geom_bar(stat = "identity") +
    coord_flip() +
    labs(x = "", y = "-log10(Adjusted P-value)", fill = "Adjusted P-value") +
    theme_classic() +
    scale_fill_gradient(low = "red", high = "blue")
  
  pdf("./image/5_gsea_recur_genes.pdf", width = 10, height = 3)
  p
  dev.off()
  
```

#------------------------------------------------------------
# 4. The protein distribution of r.v. between smk and non smk
#------------------------------------------------------------


```{r fig.width=6, fig.height=4}
library(ggplot2)
library(tidyverse)
library(biomaRt)
ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
attributes <- listAttributes(ensembl)


# 1. Load the SNVs data
  final.snps.table = readRDS("./data/1_final_SNVs_table.rds")

# 2. Rename the variants  
  final.snps.table$Func[which(final.snps.table$Func == "missense_variant"  |
                              final.snps.table$Func == "missense_variant+splice_region_variant" |
                              final.snps.table$Func == "start_lost+splice_region_variant"  |
                              final.snps.table$Func == "start_lost"|
                              final.snps.table$Func == "stop_lost" )] = "missense"
  final.snps.table$Func[which(final.snps.table$Func == "stop_gained")] = "nonsense"
  final.snps.table$Func[which(final.snps.table$Func == "splice_donor_variant+intron_variant"    |
                              final.snps.table$Func == "splice_acceptor_variant+intron_variant" |
                              final.snps.table$Func == "stop_gained+splice_region_variant")] = "splice_region"

# 3. Rename the variants      
  # muts = final.snps.table[which(final.snps.table$Gene == "PLCL2" & final.snps.table$Func == "missense"), c("Transcript","sample_id")]
  # print(table(muts$Transcript))
  # 
  # muts = final.snps.table[which(final.snps.table$Gene == "PSPH" & final.snps.table$Func == "missense"), c("Transcript","sample_id")]
  # print(table(muts$Transcript))
  # 
  # muts = final.snps.table[which(final.snps.table$Gene == "ARSD" & final.snps.table$Func == "missense"), c("Transcript","sample_id")]
  # print(table(muts$Transcript))
   final.snps.table[, c("Transcript","sample_id")]
```


```{r fig.width=6, fig.height=4}
  pdf(file = paste0("./image/aa_distr/aa_distr.pdf"), width = 6, height = 4)
  
  genes = c("MUC19", "MUC3A", "MUC5AC", "PABPC3", "SYN2")
  for (g in genes[5]) {
    g = genes[4]
    muts = final.snps.table[which(final.snps.table$Gene == g & final.snps.table$Func == "missense"), c("Transcript","sample_id")]
    protein.id = unique(sub("(.*)\\..:p\\..*/c.*", "\\1", names(table(muts$Transcript))))
    cds_info <- getBM(attributes = c("cdna_coding_start","cdna_coding_end"), filters = "refseq_mrna", values = protein.id, mart = ensembl)
    cds_length <- max(cds_info[["cdna_coding_end"]]) - min(cds_info[["cdna_coding_start"]]) + 1
    protein_length = (round(cds_length/3))
    muts.protein = sub(".*:(p\\..*)/c.*", "\\1", names(table(muts$Transcript)))
    muts.sites = gsub("[^0-9]", "", muts.protein)
    muts$protein = muts$Transcript
    muts$sites = muts$Transcript
    for (i in 1:length(table(muts$Transcript))) {
      idx = which(muts$Transcript == names(table(muts$Transcript))[i])
      muts$protein[idx] = muts.protein[i]
      muts$sites[idx] = muts.sites[i]
    }
    mutations_plot <- muts %>% 
      group_by(sites) %>% 
      summarize(count = n()) %>% 
      arrange(sites) %>% 
      mutate(order = 1:n())
    mutation_positions <- as.integer(mutations_plot$sites)
    mutation_counts <- mutations_plot$count
    amino_acid_positions <- 1:protein_length
    p = plot(amino_acid_positions, rep(0,length(amino_acid_positions)), type="n", xlab="Amino acid position", ylab="Mutation count",
             main = paste0(g), ylim=c(0,max(mutation_counts)))+
      points(mutation_positions, mutation_counts, pch=16, col="#69b3a2") +
      for(i in 1:length(mutation_positions)){
        segments(mutation_positions[i], 0, mutation_positions[i], mutation_counts[i], col="#E21818")
      }
  }

    g = "ZNF806"
    muts = final.snps.table[which(final.snps.table$Gene == g & final.snps.table$Func == "missense"), c("Transcript","sample_id")]
    protein_length = 583
    muts.protein = sub(".*:(p\\..*)/c.*", "\\1", names(table(muts$Transcript)))
    muts.sites = gsub("[^0-9]", "", muts.protein)
    muts$protein = muts$Transcript
    muts$sites = muts$Transcript
    for (i in 1:length(table(muts$Transcript))) {
      idx = which(muts$Transcript == names(table(muts$Transcript))[i])
      muts$protein[idx] = muts.protein[i]
      muts$sites[idx] = muts.sites[i]
    }
    
    mutations_plot <- muts %>% 
      group_by(sites) %>% 
      summarize(count = n()) %>% 
      arrange(sites) %>% 
      mutate(order = 1:n())
    
    mutation_positions <- as.integer(mutations_plot$sites)
    mutation_counts <- mutations_plot$count
    amino_acid_positions <- 1:protein_length
    plot(amino_acid_positions, rep(0,length(amino_acid_positions)), type="n", xlab="Amino acid position", ylab="Mutation count",
             main = paste0(g), ylim=c(0,max(mutation_counts)))+
      points(mutation_positions, mutation_counts, pch=16, col="#69b3a2") +
      for(i in 1:length(mutation_positions)){
        segments(mutation_positions[i], 0, mutation_positions[i], mutation_counts[i], col="#E21818")
      }

    dev.off()
```

```{r}
load("./data/1_final_recur_genes.RData")
recurrent_vars[grep(pattern = "^MUC", x = recurrent_vars)]
recurrent_vars
```

```{r}
pre = c("TERT", "TERC", "OBFC1", "ATP11A", "TOLLIP", "DSP", "DPP9", "AKAP", "FAM13A", "SPPL2C", "PARN", "RTEL1", "SFTPA2", "SFTPC",
        "HLA-DRB1", "HLA-DQB1", "HLA-DRB5", "HLA-A", "HLA-B", "HLA-H", "HLA-G", "HLA-DQA1", "HLA-L") 
intersect(recurrent_vars, pre)
```


```{r}
load(file = "./data/1_gsea_recur_genes.RData")
ego@result
ego@result
```


```{r}
load("./data/2_DEseq_res.RData")
data.frame(res)["MUC3A",]
data.frame(res)["DNAH11",]
```



