---
title: "SNPs data preprocessing"
author: "Shuai, Yi"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
# Add required packages
  library(ggplot2)
  library(tidyverse)
  library(hrbrthemes)
  library("stringr")
  library(ComplexHeatmap)
  library(DESeq2)
  library(org.Hs.eg.db)
  library(clusterProfiler)
  library(DOSE)
  library(plyr)
```

#------------------------------------------------------
# 1. DE analysis 
#------------------------------------------------------

```{r fig.height=5, fig.width=6}
# 1. Perform standard DE analysis
  if (file.exists("./data/2_DEseq_res.RData")) {
    message("Bulk level DE analysis, DONE!")
  } else {
    genes = read.table("./data/2_RNAseq/GSE134692_gene_annotation.txt.gz", header = T)
    row.names(genes) = genes$ensembl
    
    design = read.csv("./data/2_RNAseq/GSE134692_design.txt.gz", header = T, sep = "\t")
    row.names(design) = design$sample_id
    normal = row.names(design)[which(design$DiseaseStatus == "Normal")]
    ipf = row.names(design)[which(design$DiseaseStatus == "IPF")]
    design_fl = design[c(normal, ipf),]
    
    counts = read.table("./data/2_RNAseq/GSE134692_tmm_normalized_and_filtered_log2CPM.txt.gz", header = T)
    row.names(counts) = counts$ensembl
    counts$genename = genes[row.names(counts), "GeneName"]
    counts_fl = counts[,c("genename", normal, ipf)]
    counts_fl = ddply(counts_fl,"genename", numcolwise(sum))  
    row.names(counts_fl) = counts_fl$genename
    counts_fl_fl = round(2^counts_fl[,-1])
    # Remove un-interested genes before DE analysis
    mito.gene = grep(pattern = "^MT[ACN-]", x = row.names(counts_fl_fl))
    row.names(counts_fl_fl)[mito.gene]
    ac.gene = grep(pattern = "^A.[[:digit:]]", x = row.names(counts_fl_fl))
    row.names(counts_fl_fl)[ac.gene]
    counts_fl_fl_fl = counts_fl_fl[-c(mito.gene, ac.gene),] # 15006 genes
    # dim(counts_fl_fl_fl)
    
    # Run DEseq2
    dds <- DESeqDataSetFromMatrix(countData = (counts_fl_fl_fl[-1,]), 
                                  colData = (design_fl), 
                                  design = ~(DiseaseStatus + Batch))
    dds$DiseaseStatus <- relevel(dds$DiseaseStatus, ref="Normal")
    dds <- estimateSizeFactors(dds)
    dds <- DESeq(dds)
    res <- results(dds)
    save(res, file = "./data/2_DEseq_res.RData")
    }
  load("./data/2_DEseq_res.RData")
  
  write.csv(res[which(res$padj < 0.05 & abs(res$log2FoldChange) >= 2),], file = "./stable1_deanalysis_results.csv")
  genes = read.table("./data/2_RNAseq/GSE134692_gene_annotation.txt.gz", header = T)
  row.names(genes) = genes$ensembl
  dim(res)
  deg = res[which(res$padj < 0.05 & abs(res$log2FoldChange) >= 2),]
  dim(deg) # 876 genes
  
# 2. Volcano plot showing the DE analysis results  
  # par(mar=c(5,5,5,5), cex=1.0, cex.main=1.4, cex.axis=1.4, cex.lab=1.4)
  topT <- as.data.frame(res)
  pdf(file = "./image/6_vocalno_plots_deseq2.pdf", width = 5, height = 4)
  with(topT, plot(log2FoldChange, -log10(padj), pch=20, main="Volcano plot", cex=1.0, col = "#CFD2CF",
                  xlab=bquote(~Log[2]~fold~change), ylab=bquote(~-log[10]~adj.p-value)))
  with(subset(topT, padj < 0.05 & log2FoldChange >= 2), points(log2FoldChange, -log10(padj), pch=20, col="#EB455F", cex=1))
  with(subset(topT, padj < 0.05 & log2FoldChange <= -2), points(log2FoldChange, -log10(padj), pch=20, col="#0081C9", cex=1))
  #Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
  abline(v=0, col="black", lty=3, lwd=1.0)
  abline(v=-2, col="black", lty=2, lwd=2.0)
  abline(v=2, col="black", lty=2, lwd=2.0)
  abline(h=-log10(max(topT$pvalue[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=2.0)
  dev.off()
  
  # Basic stats of DE genes
  dim(subset(topT, padj < 0.05 & log2FoldChange <= -2))
  dim(subset(topT, padj < 0.05 & log2FoldChange >= 2))
```

```{r fig.height=5, fig.width=7}
# 3. KEGG enrichment analysis following the DE analysis results
  res$GeneSymbol = row.names(res)
  original_gene_list = res$log2FoldChange
  names(original_gene_list) <- res$GeneSymbol
  # Convert gene IDs for gseKEGG function
  ids = bitr(names(original_gene_list), fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)
  dedup_ids = ids[!duplicated(ids[c("SYMBOL")]),]
  colnames(dedup_ids) = c("GeneSymbol", "EntrezID")
  df2 = merge(data.frame(res), dedup_ids, by = "GeneSymbol")
  kegg_gene_list = df2$log2FoldChange
  names(kegg_gene_list) = df2$EntrezID
  kegg_gene_list = na.omit(kegg_gene_list)
  kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)
  kegg_organism = "hsa"
  kk2 = gseKEGG(geneList     = kegg_gene_list,
                 organism     = kegg_organism,
                 minGSSize    = 3,
                 maxGSSize    = 800,
                 pvalueCutoff = 0.05,
                 pAdjustMethod = "none",
                 keyType       = "ncbi-geneid")
  # kegg_res = subset(kk2@result, abs(NES)>1)
   Hallmark_pathway_plot <-
    ggplot(kk2@result, aes(reorder(Description, NES), NES)) + 
    geom_col(aes(fill=-log10(p.adjust))) +
    coord_flip() +
    labs(x="Pathway", y="Normalized Enrichment Score", title="Hallmark pathways NES from GSEA") + 
    theme_minimal()
   ggsave(plot = Hallmark_pathway_plot, filename = "./image/7_kegg_hallmark.pdf", dpi = 300, width = 7, height = 7) 
   write.csv(kk2@result, file = "./data/3_kegg_detailed_table.csv")
```

#---------------------------
# 2. genome vs transcriptome
#---------------------------

```{r}
intersect(row.names(deg), recurrent_vars)
```

```{r}
  deg_names = rownames(deg)
  genes_annot = genes[!duplicated(genes$GeneName),]
  row.names(genes_annot) = genes_annot$GeneName
  deg_anno = genes_annot[deg_names,]
  deg_anno$Chromosome = paste0("chr",deg_anno$Chromosome)
  chr_lenght = c(247249719, 242951149, 199501827, 191273063, 180857866, 
                 170899992, 158821424, 146274826, 140273252, 135374737, 
                 134452384, 132349534, 114142980, 106368585, 100338915, 
                 88827254,  78774742, 76117153, 63811651, 62435964,
                 46944323, 49691432, 154913754, 57772954)
  chr_lenght_rl = chr_lenght/sum(chr_lenght) * 100
  data = data.frame(deg_counts = table(deg_anno$Chromosome))
  data$counts_adj = data$deg_counts.Freq/chr_lenght_rl
  colnames(data) = c("Chr", "counts", "counts_adj")
  data$Chr = factor( data$Chr, c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9",
                                 "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17",
                                 "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"))
  p1 <-
    ggplot(data, aes(x = Chr, y = counts_adj)) +
    geom_bar(stat='identity',  fill="#69b3a2" , width=0.75) +
    ggtitle("SNPs frequencies per Chr adjusted by length") +
    theme_classic() +
    theme(
      plot.title = element_text(size=15),
      axis.text.x = element_text(angle = 45, vjust = 0.5)
    ) +
    geom_hline(yintercept = mean(data$counts_adj), linetype="dotted", color = "red", size=1)
  ggsave(plot = p1, filename = "./image/8_degenes_per_unit_chr.pdf", dpi = 300, width = 7, height = 4) 

# mutated TFs alter the expr of their target genes
# devtools::install_github("slowkow/tftargets")
  library(tftargets)
  load("./data/1_final_recur_genes.RData")
  load("./data/1_gsea_recur_genes.RData")
  load("./data/2_DEseq_res.RData")
  deg = res[which(res$padj < 0.05 & res$log2FoldChange >= 2 & res$baseMean >= 2^5),]
  deg_tf_list = NULL
  TF = intersect(names(Marbach2016), recurrent_vars) 
  # c("AR","DMRTA2","FOXD4L5","KDM6B","MEF2A","MESP2","TBP","ZIC5","HOXA1","ZNF729")
  TF
  for (i in TF) { deg_tf_list = union(deg_tf_list, intersect(Marbach2016[[i]], row.names(deg)) )  }
  length(deg_tf_list)
  write.csv(deg_tf_list,file = "./stable3_degenes_regulated_by_muts_TFs.csv")
  genes_annot = genes[!duplicated(genes$GeneName),]
  row.names(genes_annot) = genes_annot$GeneName
  genes_annot$Chromosome = paste0("chr",genes_annot$Chromosome)
  chr_lenght = c(247249719, 242951149, 199501827, 191273063, 180857866, 170899992, 158821424, 146274826, 
                 140273252, 135374737, 134452384, 132349534, 114142980, 106368585, 100338915, 88827254, 
                 78774742, 76117153, 63811651, 62435964, 46944323, 49691432, 154913754, 57772954)
  chr_lenght_rl = chr_lenght/sum(chr_lenght) * 100
  deg_sub = genes_annot[deg_tf_list,]
  deg_sub$Chromosome = factor(deg_sub$Chromosome, c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9",
                                                    "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17",
                                                    "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"))
  data = data.frame(table(deg_sub$Chromosome))
  colnames(data) = c("Chr", "counts")
  data$counts_adj = data$counts/chr_lenght_rl
  p2 <-
    ggplot(data, aes(x = Chr, y = counts_adj)) +
    geom_bar(stat='identity',  fill="#69b3a2" , width=0.75) +
    ggtitle("SNPs frequencies per Chr adjusted by length") +
    theme_classic() +
    theme(
      plot.title = element_text(size=15),
      axis.text.x = element_text(angle = 45, vjust = 0.5)
    ) +
    geom_hline(yintercept = mean(data$counts_adj), linetype="dotted", color = "red", size=1)
  ggsave(plot = p2, filename = "./image/8_degenes_byTF_per_unit_chr.pdf", dpi = 300, width = 7, height = 4) 
  # 165/876
```
#----
# End
#----

```{r fig.height=5, fig.width=6}
# Enrichment
  ego <- enrichGO(gene          = deg_names,
                  keyType       = 'SYMBOL',
                  universe      = row.names(deg),
                  OrgDb         = org.Hs.eg.db,
                  ont           = "ALL",
                  pAdjustMethod = "none", 
                  pvalueCutoff  = 0.05,
                  # qvalueCutoff  = 0.05,
                  readable      = TRUE)
  # head(ego)
  ego@result
  save(ego, file = "./data/1_gsea_recur_genes.RData")
  pdf("./image/5_gsea_recur_genes.pdf", width = 6, height = 5)
  dotplot(ego, showCategory=15)
  dev.off()
  mutate(ego, showCategory=10, qscore = -log(p.adjust, base=10)) %>%
    barplot(x="qscore")
  
```





```{r fig.height=5, fig.width=6}
# Enrichment of upreg genes
  upreg <- rownames(res)[res$padj < 0.05 & res$log2FoldChange >= 2]
  ego <- enrichGO(gene          = upreg,
                  keyType       = 'SYMBOL',
                  universe      = row.names(deg),
                  OrgDb         = org.Hs.eg.db,
                  ont           = "ALL",
                  pAdjustMethod = "none", 
                  pvalueCutoff  = 0.05,
                  # qvalueCutoff  = 0.05,
                  readable      = TRUE)
  ego <- simplify(ego)
  ego@result
  # save dotplot
  pdf("./image/7_gsea_upreg_genes.pdf", width = 5, height = 6)
  dotplot(ego, showCategory=4)
  dotplot(ego, showCategory=4, orderBy = "GeneRatio", split="ONTOLOGY")
  dev.off()
  # save ego res
  save(ego, file = "./data/3_gsea_upreg_genes.RData")
  
  # Enrichment of upreg genes
  downreg <- rownames(res)[res$padj < 0.05 & res$log2FoldChange <= -2]
  ego <- enrichGO(gene          = downreg,
                  keyType       = 'SYMBOL',
                  universe      = row.names(deg),
                  OrgDb         = org.Hs.eg.db,
                  ont           = "ALL",
                  pAdjustMethod = "none", 
                  pvalueCutoff  = 0.05,
                  # qvalueCutoff  = 0.05,
                  readable      = TRUE)
  ego <- simplify(ego)
  ego@result
  # save dotplot
  pdf("./image/7_gsea_downreg_genes.pdf", width = 5, height = 4)
  dotplot(ego, showCategory=4, orderBy = "GeneRatio", split="ONTOLOGY")
  dev.off()
  # save ego res
  save(ego, file = "./data/3_gsea_downreg_genes.RData")
```



